### DISCUSSION TITLE
Home Assistant MQTT discovery still occurring with SetOption19=Off (Treatlife DS01 dimmer)

### DISCUSSION THREAD

**Discussion by mcarson75** [arendst/Tasmota#16133] [Q&A] [answered]
I have been using a mix of Tasmota discovery and Home Assistant discovery.  However, in preparation to upgrade to Tasmota 12.0.2, I have been moving all of my devices over to Tasmota discovery (`SO19 0`).  In most cases this has worked flawlessly.  On my Treatlife DS01 dimmers, though, when I disable SO19 the Tasmota discovery works as expected:

![image](https://user-images.githubusercontent.com/54790846/182263706-7721b464-ff13-41fe-81d2-ccb25a5e27a2.png)

However, it is still also creating a device under the MQTT broker in HA, as shown:

![image](https://user-images.githubusercontent.com/54790846/182263803-58c45215-6198-4b9c-b620-28dc4ca6f277.png)

I have tried deleting both the MQTT device in Home Assistant and the accompanying topic in the MQTT broker, but it reappears when I restart the device.

This appears under the topic `homeassistant/light/45F476_LI_1/config`, with the following payload:

```
{
  "platform": "mqtt",
  "name": "Front Path Light",
  "stat_t": "tasmota/outside/front-path/light/stat/RESULT",
  "stat_val_tpl": "{{ value_json.POWER }}",
  "cmd_t": "tasmota/outside/front-path/light/cmnd/POWER",
  "avty_t": "tasmota/outside/front-path/light/tele/LWT",
  "bri_stat_t": "tasmota/outside/front-path/light/stat/RESULT",
  "bri_cmd_t": "tasmota/outside/front-path/light/cmnd/Dimmer",
  "bri_scl": 100,
  "bri_val_tpl": "{{ value_json.Dimmer }}",
  "on_cmd_type": "brightness",
  "pl_off": "OFF",
  "pl_on": "ON",
  "pl_avail": "Online",
  "pl_not_avail": "Offline",
  "ret": false,
  "uniq_id": "45F476_LI_1",
  "dev": {
    "ids": [
      "45F476"
    ]
  }
}
```

Is this expected?  If not, how can I fix it?

EDIT to add: This happens both in Tasmota 11.0.0 and 12.0.2

**Comment by barbudor** [comment]
Can you check with an mqtt logger what the dimmer is publishing on both 'homeasistant/#' and 'tasmota/#' discovery topics?
If it confirms that Tasmota is still publishing something on 'homeassistant' while so19 is 0 then the problem is in Tasmota.

However your edited comment is unexpected
Are you using a Tasmota official binary?
As this do not include anymore HA code it shouldn't publish on 'homeassistant'

**Reply by sfromis** [reply]
You likely have some old retained messages lingering on the MQTT Discovery topics. Normally, Tasmota will try to clean those up when switching, but probably not if installing Tasmota version 12 without the deprecated MQTT discovery code present.

**Reply by mcarson75** [reply]
I did an OTA upgrade from [http://ota.tasmota.com/tasmota/release/tasmota.bin.gz](http://ota.tasmota.com/tasmota/release/tasmota.bin.gz), and it is on the official Tasmota 12.0.2.

![image](https://user-images.githubusercontent.com/54790846/182384847-7062b18d-cb2e-4ffe-a0ec-65e871c961bd.png)

Here are the topics under both homeassistant and tasmota as shown in MQTT Explorer:
![image](https://user-images.githubusercontent.com/54790846/182385061-f2cfdeed-ea27-406d-92cf-a00183cc5bb6.png)

To test I deleted both devices (Tasmota and MQTT) from Home Assistant, and deleted the retained topics via MQTT Explorer:
![image](https://user-images.githubusercontent.com/54790846/182385969-d51b1ac7-1969-4cb7-ab09-f4475b22802c.png)

Then I restarted the device, and both topics reappear:
![image](https://user-images.githubusercontent.com/54790846/182386343-1ba83d11-7767-4b4a-ab2f-6a9be061d061.png)

Both devices reappear in Home Assistant.

I should note that with SO19 off, it only has the MQTT message under "light" and not the status message under "sensor" (here is a shot from an identical dimmer with Tasmota 11.0.0 and SO19 on).  This is why it shows as `<unknown>` for Manufacturer and Model:
![image](https://user-images.githubusercontent.com/54790846/182387382-de9b0b77-e931-41e2-874a-3fe2accf9ea8.png)


And just to verify that SetOption19 is off - 
![image](https://user-images.githubusercontent.com/54790846/182386522-be3dfaec-7ae0-4bf2-a82a-5f2b1196e44e.png)

**Reply by barbudor** [reply]
Thanks,
I'm having a look

**Comment by barbudor** [comment, marked as answer]
I can't reproduce your issue.
I have loaded a NodeMCU with latest Tasmota Development v12.0.2.4 from today
Connected to a fresh test MQTT broker (only this device on it, started from scratch)
Configured template for Treatlife DS01 from the template repositry
Here is what I see on the MQTT broker when I boot the device:
```
tele/tasmota_A6D683/LWT Online
cmnd/tasmota_A6D683/POWER (null)
tele/tasmota_A6D683/INFO1 {"Info1":{"Module":"DS02S Dimmer","Version":"12.0.2.4(tasmota)","FallbackTopic":"cmnd/DVES_A6D683_fb/","GroupTopic":"cmnd/tasmotas/"}}
tele/tasmota_A6D683/INFO2 {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-A6D683-5763","IPAddress":"192.168.1.103"}}
tele/tasmota_A6D683/INFO3 {"Info3":{"RestartReason":"Software/System restart","BootCount":5}}
stat/tasmota_A6D683/RESULT {"POWER":"OFF"}
stat/tasmota_A6D683/POWER OFF
tele/tasmota_A6D683/STATE {"Time":"2022-08-02T20:48:24","Uptime":"0T00:00:09","UptimeSec":9,"Heap":27,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":25,"MqttCount":1,"POWER":"OFF","Wifi":{"AP":1,"SSId":"mado-wifi","BSSId":"E8:1B:69:02:D4:AA","Channel":1,"Mode":"11n","RSSI":44,"Signal":-78,"LinkCount":1,"Downtime":"0T00:00:03"}}
tasmota/discovery/18FE34A6D683/config {"ip":"192.168.1.103","dn":"Tasmota","fn":["Tasmota",null,null,null,null,null,null,null],"hn":"tasmota-A6D683-5763","mac":"18FE34A6D683","md":"DS02S Dimmer","ty":1,"if":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"12.0.2.4","t":"tasmota_A6D683","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[1,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null],"btn":[0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":0,"lt_st":0,"sho":[0,0,0,0],"sht":[[0,0,0],[0,0,0],[0,0,0],[0,0,0]],"ver":1}
tasmota/discovery/18FE34A6D683/sensors {"sn":{"Time":"2022-08-02T20:48:29"},"ver":1}
```

So no publication on homeassitant topics, only tasmota/discovery
As @sfromis I really bet for a retained topic

**Reply by mcarson75** [reply]
I don't doubt that it's something like that, but I'm wondering how I would find that?  As mentioned, I deleted the retained topic in /homeassistant and it deletes the MQTT device.  However, this topic returns when I reboot the device.  Where else would I look for a retained topic?  I can send my device config also if it is useful.

**Reply by barbudor** [reply]
You should be able to detele retained messages with MQTT Explorer
One easy way to confirm it is retained and not sent again and again by Tasmota is to
- shutdown the Tasmota device
- start MQTT Explorer
- delete retained messages
- quit and restart MQTT explorer
=> if the messsages are back again while you haven't restarted Tasmota, it means that they are still floating around
- then boot your Tasmota
=> Only the tasmota/discovery should pop up

To properly delete a retained message in MQTT Explorer, you should use the trash icon or the "Del" key. MQTT explorer will ask for confirmation of deletion
<img width="680" alt="image" src="https://user-images.githubusercontent.com/2996042/182472257-141aa91b-5174-4289-b64e-67b5e452fdf1.png">

**Reply by mcarson75** [reply]
Yes, it absolutely comes back if I follow that procedure.  I videoed the process so you can see - [https://www.youtube.com/watch?v=zD44ivtJLUc](https://www.youtube.com/watch?v=zD44ivtJLUc)

**Reply by mcarson75** [reply]
Ha, I just figured it out.  This was in my rule1.  No idea why.

```
{"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":672,"Free":131,"Rules":"on system#boot do backlog delay 30; publish2 homeassistant/light/45F476_LI_1/config {\"platform\":\"mqtt\",\"name\":\"Front Path Light\",\"stat_t\":\"tasmota/outside/front-path/light/stat/RESULT\",\"stat_val_tpl\":\"{{ value_json.POWER }}\",\"cmd_t\":\"tasmota/outside/front-path/light/cmnd/POWER\",\"avty_t\":\"tasmota/outside/front-path/light/tele/LWT\",\"bri_stat_t\":\"tasmota/outside/front-path/light/stat/RESULT\",\"bri_cmd_t\":\"tasmota/outside/front-path/light/cmnd/Dimmer\",\"bri_scl\":100,\"bri_val_tpl\":\"{{ value_json.Dimmer }}\",\"on_cmd_type\":\"brightness\",\"pl_off\":\"OFF\",\"pl_on\":\"ON\",\"pl_avail\":\"Online\",\"pl_not_avail\":\"Offline\",\"ret\":false,\"uniq_id\":\"45F476_LI_1\",\"dev\":{\"ids\":[\"45F476\"]}} endon"}}
```

**Reply by barbudor** [reply]
Great to know you found it

Also in your video at 1:19, you are closing MQTT Explorer *without* having deleted the homeassitant topic

**Reply by mcarson75** [reply]
I have found some issues with MQTT Explorer still displaying a deleted topic when you have something in the search bar.  I've seen it before and if you clear the search bar it shows correctly, although I didn't do it in this video.  I should have done it for completeness, but it was gone when I restarted MQTT Explorer.
